### PR TITLE
#53837 Move the PHPCS basepath config out of phpcs.xml.dist

### PR DESCRIPTION
**Pull request by johnbillion**
Trac ticket: https://core.trac.wordpress.org/ticket/53837

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -35,7 +35,7 @@
 	"scripts": {
 		"compat": "@php ./vendor/squizlabs/php_codesniffer/bin/phpcs --standard=phpcompat.xml.dist --report=summary,source",
 		"format": "@php ./vendor/squizlabs/php_codesniffer/bin/phpcbf --report=summary,source",
-		"lint": "@php ./vendor/squizlabs/php_codesniffer/bin/phpcs --report=summary,source",
+		"lint": "@php ./vendor/squizlabs/php_codesniffer/bin/phpcs --report=summary,source --basepath='./'",
```
Comment (jrfnl) [inline composer.json:38]:
```suggestion
		"lint": "@php ./vendor/squizlabs/php_codesniffer/bin/phpcs --report=summary,source --basepath=./",
```

The quotes shouldn't be needed.

**Review by jrfnl — COMMENTED** [review]
Generally speaking, I'm not a fan of this change as it is a work-around for a select group of users using a specific tool, which makes things less convenient for the majority of users. To me, that's the world upside down.

I don't use VS Code myself, so I haven't checked if there isn't some way it can be changed in the configuration for PHPCS in VS Code.

On the command-line, the `basepath` setting in the ruleset can be overruled by passing `--basepath=` (without value) and you'd get the full path to the files in the report, so if you can add command line parameters to the PHPCS command in VS Code, that should be the preferred solution.

The other side of the argument is, that the `basepath` argument cannot easily be overloaded via a `[.]phpcs.xml` overload file. (Expected to be changed in PHPCS 4.x, but that's still a while off)

So, while **without** this PR, users of VS Code can only get the full path if they can pass the `--basepath=` argument to the command used in VS Code.

While **with** this PR, users who prefer clean paths have three options:
1. Use PHPCS via the Composer script.
2. Pass `--basepath=.` themselves on the command-line.
3. Use a `[.]phpcs.xml` overload file, like so:
    ```xml
	<ruleset name="WP Core local">
		<arg name="basepath" value="./"/>
		<rule ref="./phpcs.xml.dist"/>
	</rule>
    ```

But the fact that these options are available, is largely unknown to users, so we'd still be inconveniencing the majority of users in support of the minority.

More than anything, I believe the Make Core handbook should probably be enhanced with a few additional pages, informing users about how to install and run PHPCS (and related style tooling for CSS and JS) and with answers to commonly asked questions about these tools and links to their manuals/external tutorials on how to set things up in combination with IDEs.

Code-wise the PR looks okay, just one tiny remark.

However, I do think it needs checking that this change does not impact the `cs2pr` integration in GH Actions.
This can be tested by introducing some errors in a second commit on this PR and seeing if those show up inline on the "Files" tab (as expected).

**Comment by johnbillion** [comment]
> it is a work-around for a select group of users using a specific tool, which makes things less convenient for the majority of users

My experience says this isn't the case, due to the way people most commonly use PHPCS. I work with various teams of experienced and inexperienced developers and almost none of them run PHPCS directly using `./vendor/bin/phpcs` or a global `phpcs`, the majority of them interface with it as just another tool in their editor/IDE or via a project-level Composer or NPM script. VS Code has 14 million users (!) and the PHPCS extension over a million, so I'm sure this change will convenience many more users than those who are familiar with and use `phpcs` directly. Plus it's an immediate fix for VS Code users who want to contribute to WordPress, which is the main thing.

> I believe the Make Core handbook should probably be enhanced with a few additional pages, informing users about how to install and run PHPCS

Yeah agreed. Maybe one for the docs team and test team to collaborate on.

> I do think it needs checking that this change does not impact the `cs2pr` integration in GH Actions

I don't _think_ it will impact it because the usage example given on https://github.com/staabm/annotate-pull-request-from-checkstyle uses a vanilla PHPCS configuration without any value in `basepath`, but I'll check to make sure. Thanks.

**Comment by jrfnl** [comment]
> 
> 
> > it is a work-around for a select group of users using a specific tool, which makes things less convenient for the majority of users
> 
> My experience says this isn't the case, due to the way people most commonly use PHPCS. I work with various teams of experienced and inexperienced developers and almost none of them run PHPCS directly using `./vendor/bin/phpcs` or a global `phpcs`, the majority of them interface with it as just another tool in their editor/IDE or via a project-level Composer or NPM script. VS Code has 14 million users (!) and the PHPCS extension over a million, so I'm sure this change will convenience many more users than those who are familiar with and use `phpcs` directly. Plus it's an immediate fix for VS Code users who want to contribute to WordPress, which is the main thing.


But we're not talking about all VS Code users here, we're talking about WP Core contributors. And while I don't have access to stats, I know a wide variety of IDEs and code editors is used by those ([PHPStorm, Sublime, Atom, Eclipse, GH Code spaces and yes, VS Code](https://github.com/WordPress/WordPress-Coding-Standards/wiki), to name a few) and this change is only for the users of VS Code and may or may not negatively impact users of other editors.

Have any tests been done to see if this change impacts users of other editors ?

**Comment by johnbillion** [comment]
The fix is to revert a non-default PHPCS setting to its default; what negative impact in other editors are you imagining?

**Comment by jrfnl** [comment]
> 
> 
> The fix is to revert a non-default PHPCS setting to its default; what negative impact in other editors are you imagining?

Not imagining anything, just wanted to ask the question as if we're making changes to accommodate one editor, it seemed sensible to make sure there is no impact on other editors.

**Pull request closed without merging**